public/lesswrong-editor/src/ckeditor5-math/utils.js

```diff
@@ -0,0 +1,106 @@
+/* eslint-disable max-len */
+/* eslint-disable no-tabs */
+/* globals MathJax */
+
+export const defaultConfig = {
+	engine: 'mathjax',
+	outputType: 'script',
+	forceOutputType: false,
+	enablePreview: true
+};
+
+export function getSelectedMathModelWidget( selection ) {
+	const selectedElement = selection.getSelectedElement();
+	if ( selectedElement && ( selectedElement.is( 'mathtex' ) || selectedElement.is( 'mathtex-display' ) ) ) {
+		return selectedElement;
+	}
+	return null;
+}
+
+// Simple MathJax 3 version check
+export function isMathJaxVersion3( version ) {
+	return version && typeof version === 'string' && version.split( '.' ).length === 3 && version.split( '.' )[ 0 ] === '3';
+}
+
+// Check if equation has delimiters
+export function hasDelimiters( text ) {
+	return text.match( /^(\\\[.*?\\\]|\\\(.*?\\\))$/ );
+}
+
+// Find delimiters count
+export function delimitersCounts( text ) {
+	return text.match( /(\\\[|\\\]|\\\(|\\\))/g ).length;
+}
+
+// Extract delimiters and figure display mode for the model
+export function extractDelimiters( equation ) {
+	equation = equation.trim();
+
+	// Remove delimiters (e.g. \( \) or \[ \])
+	const hasInlineDelimiters = equation.includes( '\\(' ) && equation.includes( '\\)' );
+	const hasDisplayDelimiters = equation.includes( '\\[' ) && equation.includes( '\\]' );
+	if ( hasInlineDelimiters || hasDisplayDelimiters ) {
+		equation = equation.substring( 2, equation.length - 2 ).trim();
+	}
+
+	return {
+		equation,
+		display: hasDisplayDelimiters
+	};
+}
+
+export async function renderEquation( equation, element, engine = 'mathjax', display = false, preview = false, previewUid ) {
+	if ( engine === 'mathjax' && typeof MathJax !== 'undefined' ) {
+		if ( isMathJaxVersion3( MathJax.version ) ) {
+			await renderMathJax3( equation, element, display, preview );
+		}
+	} else {
+		element.innerHTML = equation;
```

Comment:
This line is really scary, and my intuition is that there is probably a cross-site scripting vulnerability here.

Reply:
Yeah, seems like we should just change it to `innerText`. Will go and try that on Monday.